### PR TITLE
Close: review readers

### PR DESCRIPTION
**Pull request by jburel**
Make sure resources are always closed
Check that the tests remain green
https://trello.com/c/NEWizoBO/90-tiffparser-leaking-resource

### CODE REVIEW

File: components/formats-gpl/src/loci/formats/in/FV1000Reader.java [hunk 1]
```diff
@@ -180,14 +180,19 @@ public int getOptimalTileWidth() {
...
       IFD ifd = tp.getFirstIFD();
-      plane.close();
       return (int) ifd.getTileWidth();
     }
     catch (FormatException e) {
       LOGGER.debug("Could not retrieve tile width", e);
     }
     catch (IOException e) {
       LOGGER.debug("Could not retrieve tile width", e);
+    } finally {
```
Comment (joshmoore) [inline components/formats-gpl/src/loci/formats/in/FV1000Reader.java]: You can use `RAIS plane2 = plane` here, e.g.:

```
import java.io.Closeable;

class C implements Closeable {
    public void close() {
        System.out.println("closed");
    }
}

public class M {
    public static void main(String[] args) {
        C c = new C();
        try (C c2 = c) {
            System.out.println("done");
        }
    }
}
```

**Comment by jburel** [comment]
@joshmoore updated

**Comment by jburel** [comment]
As discussed today, there is a commit per file modified

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [BIOFORMATS-DEV-merge-push#1593](https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-push/1593/). See the [console output](https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-push/1593/consoleText) for more details.
Possible conflicts:
  - PR #3261 melissalinkert 'Fixes for writing TIFF files in memory'
    - components/formats-bsd/src/loci/formats/out/TiffWriter.java

--conflicts

**Comment by joshmoore** [comment]
As a side-node: for merging, I'd propose that we use the `Squash and merge` functionality, otherwise someone following the git history will be confronted with an ordeal.

File: components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java [hunk 2]
```diff
@@ -167,8 +167,6 @@ public void close() throws IOException {
           saver.overwriteIFDOffset(in, allOffsets[mainIFDIndex], nextPointer);
           saver.overwriteIFDValue(in, currentFullResolution, IFD.SUB_IFD, subIFDOffsets);
           saver.close();
-          out.close();
-          in.close();
```
Comment (dgault) [inline components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: Is it correct to remove these 2 close statments?

Comment (jburel) [reply components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: yes since they are closed in the ``finally``

Comment (joshmoore) [reply components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: @jburel, except the `in` and `out` variable get overwritten on lines 158 and 153, respectively, meaning only the _last_ file in the loop will be closed.

Comment (jburel) [reply components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: correct too much cleaning. I will fix that

File: components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java [hunk 3]
```diff
@@ -167,6 +167,7 @@ public void close() throws IOException {
           saver.overwriteIFDOffset(in, allOffsets[mainIFDIndex], nextPointer);
           saver.overwriteIFDValue(in, currentFullResolution, IFD.SUB_IFD, subIFDOffsets);
           saver.close();
+          in.close();
```
Comment (joshmoore) [inline components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: I'd be for this being in a try-finally or try-resource block.

Comment (jburel) [reply components/formats-bsd/src/loci/formats/out/PyramidOMETiffWriter.java]: The full block needs to be reviewed, I was trying to keep it "simple" but it is not going to work
I will get back to it

File: components/formats-gpl/src/loci/formats/in/TargaReader.java [hunk 4]
```diff
@@ -85,12 +85,11 @@ public TargaReader() {
     options.maxBytes = planeSize;
     options.bitsPerSample = bits;
 
-    RandomAccessInputStream s = in;
-
     if (compressed) {
```
Comment (dgault) [inline components/formats-gpl/src/loci/formats/in/TargaReader.java]: The changes to this reader are resulting in test failures for testPixelsHashes. See https://web-proxy.openmicroscopy.org/east-ci/job/BIOFORMATS-test-folder/43281/console

**Comment by jburel** [comment]
Should I relist for review?

**Comment by dgault** [comment]
I have been able to review each of the changes and run all of the unit and repo tests. There is no longer any red in the tests so this PR is good to merge. I shall squash the commits as recommended.

**Pull request merged**